web/ww.ts

```diff
@@ -135,215 +270,106 @@ class Wormhole {
-					this.fail("could not generate key");
-					return;
-				}
-				console.log("generated key");
-				this.ws.send(msgB);
-				this.state = "wait_for_pc_initialize";
-				this.promise2.then(async () => {
-					if (!this.key || !this.pc) { return }
-
-					const offer = await this.pc.createOffer();
-					console.log("created offer");
-					this.ws.send(webwormhole.seal(this.key, JSON.stringify(offer)));
-					this.state = "wait_for_webtc_answer";
-					this.pc.setLocalDescription(offer);
-				});
+	makePeerConnection(iceServers: GoICEServers): RTCPeerConnection {
+		let normalisedICEServers = [];
+		for (let i = 0; i < iceServers.length; i++) {
+			normalisedICEServers.push({
+				urls: iceServers[i].URLs,
+				username: iceServers[i].Username,
+				credential: iceServers[i].Credential,
+			});
+		}
+		const pc = new RTCPeerConnection({
+			iceServers: normalisedICEServers,
+		});
+		pc.onicecandidate = (e: RTCPeerConnectionIceEvent) => {
+			if (!this.ws || !this.key || !this.pc) {
 				return;
 			}
 
-			case "wait_for_pake_b": {
-				console.log("got pake message b:", m.data);
-				this.key = webwormhole.finish(m.data);
-				if (!this.key) {
-					this.fail("could not generate key");
-					return;
-				}
-				console.log("generated key");
-				this.state = "wait_for_webtc_offer";
-				return;
+			if (e.candidate && e.candidate.candidate !== "") {
+				console.log("got local candidate", e.candidate.candidate);
+				this.ws.send(webwormhole.seal(this.key, JSON.stringify(e.candidate)));
 			}
+		};
+		return pc;
+	}
 
-			case "wait_for_webtc_offer": {
-				if (!this.key || !this.pc) { return }
+	async connType(): Promise<string> {
+		if (!this.pc) {
+			return "";
+		}
 
-				const msg: RTCSessionDescriptionInit | null = JSON.parse(webwormhole.open(this.key, m.data));
-				if (!msg) {
-					this.fail("bad key");
-					this.ws.send(webwormhole.seal(this.key, "bye"));
-					this.ws.close(WormholeErrorCodes.closeBadKey);
-					return;
-				}
-				if (msg.type !== "offer") {
-					console.log("unexpected message", msg);
-					this.fail("unexpected message");
-					return;
-				}
-				console.log("got offer");
-				// No intermediate state wait_for_pc_initialize because candidates can
-				// staring arriving straight after the offer is sent.
-				this.state = "wait_for_candidates";
-				this.promise2.then(async () => {
-					if (!this.key || !this.pc) { return }
-
-					await this.pc.setRemoteDescription(new RTCSessionDescription(msg));
-					const answer = await this.pc.createAnswer();
-					console.log("created answer");
-					this.ws.send(webwormhole.seal(this.key, JSON.stringify(answer)));
-					this.resolve3(webwormhole.fingerprint(this.key));
-					this.pc.setLocalDescription(answer);
-				});
-				return;
+		// RTCStatsReport.forEach is all that's defined in the TypeScript DOM defs, which
+		// makes this kind of awkward. Ah well.
+		const stats = await this.pc.getStats();
+
+		let id: string | undefined;
+		stats.forEach((s) => {
```

Comment:
slightly nicer, though it is a shame TS is not up to date. (Presumably, it's just TS lagging?)
```suggestion
		stats.forEach((s: RTCIceCandidatePairStats) => {
```

Reply:
possibly just TS types lagging, but also the stats report bits of WebRTC are a bit inconsistent between browsers so i'm not surprised they're not included